File: tp-5/Set.hs
module Set
    (Set, emptyS, addS, belongs, sizeS, removeS, unionS, setToList)
where

data Set a = TRS [a] Int
                  -- Cantidad de Elementos en [a].
    deriving Show
{- INV. REP.:
    * Sea (TRS xs n): no hay elementos repetidos en xs.
    * Sea (TRS xs n): n equivale a la cantidad de elementos en xs.
-}

{- COSTO OPERACIONAL DE CADA FUNCIÓN:

- emptyS        O(1)
- addS          O(n^2)
- belongs       O(n)
- sizeS         O(1)
- removeS       O(n^2)
- unionS        O(n^2)
- setToList     O(1)

-}

emptyS :: Set a
-- PROP: Crea un conjunto vacío.
emptyS = TRS [] 0


addS :: Eq a => a -> Set a -> Set a
-- PROP: Dados un elemento y un conjunto, agrega el elemento al conjunto.
addS x (TRS xs n) = if perteneceA x xs
                       then TRS xs n
                       else TRS (x:xs) (n+1)

perteneceA :: Eq a => a -> [a] -> Bool
perteneceA x []     = False
perteneceA x (y:ys) = x == y || perteneceA x ys


belongs :: Eq a => a -> Set a -> Bool
-- PROP: Dados un elemento y un conjunto indica si el elemento pertenece al conjunto.
belongs x (TRS xs n) = perteneceA x xs 


sizeS :: Eq a => Set a -> Int
-- PROP: Devuelve la cantidad de elementos distintos de un conjunto.
sizeS (TRS xs n) = n


removeS :: Eq a => a -> Set a -> Set a
-- PROP: Borra un elemento del conjunto.
removeS x (TRS xs n) = TRS (removeX x xs) (length (removeX x xs))

removeX :: Eq a => a -> [a] -> [a]
removeX x []     = []
removeX x (y:ys) = if x==y
                      then ys
                      else y : removeX x ys 


unionS :: Eq a => Set a -> Set a -> Set a
-- PROP: Dados dos conjuntos devuelve un conjunto con todos los elementos de ambos conjuntos.
unionS (TRS xs n) (TRS ys m) = TRS (unionDe xs ys) (length (unionDe xs ys))

unionDe :: Eq a => [a] -> [a] -> [a]
unionDe []     ys     = ys
unionDe xs     []     = xs
unionDe (x:xs) (y:ys) = if not (perteneceA x (unionDe xs ys))
                           then x : unionDe xs ys
                           else unionDe xs ys


setToList :: Eq a => Set a -> [a]
-- PROP: Dado un conjunto devuelve una lista con todos los elementos distintos del conjunto.
setToList (TRS xs n) = xs

File: tp-11/Set.cpp
#include <iostream>
#include "Set.h"
using namespace std;

Set emptyS(){
// PROPÓSITO: Crea un conjunto vacío.
// COSTO: O(1).
// Siendo de costo constante ya que solamente se modificaron los campos del Set dado.
    SetSt* s = new SetSt();
    s->cantidad = 0;
    s->primero = NULL;
    return s;
}

bool isEmptyS(Set s){
// PROPÓSITO: Indica si el conjunto dado está vacío.
// COSTO: O(1).
// Siendo de costo constante ya que solamente se modificaron los campos del Set dado.
    return s->cantidad == 0;
}

bool belongsS(int x, Set s){
// PROPÓSITO: Indica si el elemento pertenece al conjunto.
// COSTO: O(N).
// Siendo N la cantidad de elementos diferentes en el Set...
    NodoS* actual = s->primero;

    while (actual != NULL) {
        if (actual->elem == x) {
            break;
        }
        actual = actual->siguiente;
    }

    return actual != NULL && actual->elem == x;
}

void AddS(int x, Set s){
// PROPÓSITO: Agrega el elemento dado al conjunto dado.
// COSTO: O(N).
// Siendo N la cantidad de elementos diferentes en el Set, por cada N se realizan operaciones de costo constante. Es por
// eso que el costo total de la función es lineal.
    NodoS* actual = s->primero;

    while (actual != NULL) {
        if (actual->elem == x) {
            return;
        }
        actual = actual->siguiente;
    }

    NodoS* nuevo = new NodoS();
    nuevo->elem = x;
    nuevo->siguiente = s->primero;

    s->primero = nuevo;
    s->cantidad++;
}

void RemoveS(int x, Set s){
// PROPÓSITO: Quita el elemento dado del conjunto dado.
// PRECONDICIÓN: El Set dado no está vacío.
// COSTO: O(N).
// Siendo N la cantidad de elementos diferentes en el Set, por cada N se realizan operaciones de costo constante. Es por
// eso que el costo total de la función es lineal.
    NodoS* actual = s->primero;

    if (actual != NULL && actual->elem == x){
        NodoS* aBorrar = actual;
        s->primero = actual->siguiente;
        delete aBorrar;
        s->cantidad--;
        return;
    }

    while (actual != NULL && actual->siguiente != NULL) {
        if (actual->siguiente->elem == x) {
            NodoS* aBorrar = actual->siguiente;
            actual->siguiente = aBorrar->siguiente;
            delete aBorrar;
            s->cantidad--;
            return;
        }
        actual = actual->siguiente;
    }
}

int sizeS(Set s){
// PROPÓSITO: Devuelve la cantidad de elementos del conjunto dado.
// COSTO: O(1).
// Siendo de costo constante ya que solamente se devuelve uno de los campos del Set dado.
    return s->cantidad;
}

LinkedList setToList(Set s){
// PROPÓSITO: Devuelve una lista con los elementos del conjunto dado.
// COSTO: O(N).
// Siendo N la cantidad de elementos diferentes en el Set, por cada N se realizan operaciones de costo constante. Es por
// eso que el costo total de la función es lineal.
    NodoS* actual = s->primero;
    LinkedList xs = nil();

    while (actual != NULL) {
        Cons(actual->elem, xs);
        actual = actual->siguiente;
    }
    
    return xs;
}

void DestroyS(Set s){
// PROPÓSITO: Libera la memoria ocupada por el conjunto dado.
// COSTO: O(N).
// Siendo N la cantidad de elementos diferentes en el Set, por cada N se realizan operaciones de costo constante. Es por
// eso que el costo total de la función es lineal.
    NodoS* actual = s->primero;

    while (actual != NULL) {
        NodoS* aBorrar = actual;
        actual = actual->siguiente;
        delete aBorrar;
    }

    delete s;
}